proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java

```diff
@@ -325,34 +329,45 @@ public ProctorResult determineTestGroups(
         }
 
         final Map<String, ValueExpression> localContext =
-                ProctorUtils.convertToValueExpressionMap(
+                ProctorUtils.convertLocalContextToValueExpressionMap(
                         RuleEvaluator.EXPRESSION_FACTORY, inputContext);
 
         for (final String testName : filteredEvaluationOrder) {
             final TestChooser<?> testChooser = testChoosers.get(testName);
             final String identifier;
+            final boolean containsUnitlessAllocations =
+                    testTypesWithInvalidIdentifier.contains(
+                                    testChooser.getTestDefinition().getTestType())
+                            && testChooser.getTestDefinition().getContainsUnitlessAllocation();
+            if (!containsUnitlessAllocations) {
+                localContext.put(
+                        UNITLESS_ALLOCATION_IDENTIFIER,
+                        RuleEvaluator.EXPRESSION_FACTORY.createValueExpression(
+                                false, Object.class));
+            }
             if (testChooser instanceof StandardTestChooser) {
                 final TestType testType = testChooser.getTestDefinition().getTestType();
-                if (testTypesWithInvalidIdentifier.contains(testType)) {
+                if (testTypesWithInvalidIdentifier.contains(testType)
+                        && !containsUnitlessAllocations) {
                     // skipping here to make it use the fallback bucket.
                     continue;
                 }
 
                 identifier = identifiers.getIdentifier(testType);
-                if (identifier == null) {
+                if (identifier == null && !containsUnitlessAllocations) {
                     // No identifier for the testType of this chooser, nothing to do
                     continue;
                 }
             } else {
-                if (!identifiers.isRandomEnabled()) {
+                if (!identifiers.isRandomEnabled() && !containsUnitlessAllocations) {
                     // test wants random chooser, but client disabled random, nothing to do
                     continue;
                 }
                 identifier = null;
             }
 
             final TestChooser.Result chooseResult;
-            if (identifier == null) {
+            if (identifier == null && !containsUnitlessAllocations) {
```

Comment:
what will be the behavior for identifier == null && containsUnitlessAllocations?

Reply:
if it is null and contains unitless allocation is uses the standard test chooser to correctly bucket to the allocation with the missingExperimentalUnit rule or fallback if no match.

Reply:
why do we let random test types fall into standard chooser? Although it should not make difference.